linkerd/app/inbound/src/http/server.rs

```diff
@@ -83,3 +85,54 @@ impl<H> Inbound<H> {
         })
     }
 }
+
+// === impl ServerRescue ===
+
+impl ServerRescue {
+    /// Synthesizes responses for HTTP requests that encounter proxy errors.
+    pub fn layer() -> errors::respond::Layer<Self> {
+        errors::respond::NewRespond::layer(Self)
+    }
+}
+
+impl errors::HttpRescue<Error> for ServerRescue {
+    fn rescue(&self, error: Error) -> Result<errors::SyntheticHttpResponse> {
+        if Self::has_cause::<crate::policy::DeniedUnauthorized>(&*error) {
+            return Ok(errors::SyntheticHttpResponse {
+                http_status: http::StatusCode::FORBIDDEN,
+                grpc_status: errors::Grpc::PermissionDenied,
+                close_connection: true,
+                message: error.to_string(),
+            });
+        }
+
+        if Self::has_cause::<crate::GatewayDomainInvalid>(&*error) {
+            return Ok(errors::SyntheticHttpResponse {
+                http_status: http::StatusCode::BAD_REQUEST,
+                grpc_status: errors::Grpc::InvalidArgument,
+                close_connection: true,
+                message: error.to_string(),
+            });
+        }
+
+        if Self::has_cause::<crate::GatewayIdentityRequired>(&*error) {
+            return Ok(errors::SyntheticHttpResponse {
+                http_status: http::StatusCode::FORBIDDEN,
+                grpc_status: errors::Grpc::Unauthenticated,
+                close_connection: true,
+                message: error.to_string(),
+            });
+        }
+
+        if Self::has_cause::<crate::GatewayLoop>(&*error) {
+            return Ok(errors::SyntheticHttpResponse {
+                http_status: http::StatusCode::LOOP_DETECTED,
+                grpc_status: errors::Grpc::Aborted,
+                close_connection: true,
+                message: error.to_string(),
+            });
```

Comment:
...okay, this will recurse over the whole cause chain as much as 4 times. i wonder if maybe what we should do instead is have something like

```rust
fn traverse_causes<T>(error: &(dyn Error + 'static), f: impl FnMut(&(dyn Error + 'static)) -> Option<T>) -> Option<T> {
     f(error).or_else(error.source().map(|s| traverse_causes(s, f)))
}
```
so that we can pass in an error that tries to downcast to all the error types on each cause? that way we only recurse once.

but, this may not be that big a deal...

Reply:
Yeah, I'm aware that this isn't ideal, but it's exceptional by definition: we only expect to hit this when the proxy hits a runtime error; and, conservatively, the max chain depth we're going to see is like 10 (probably much lower than that in practice). So O(40) vs O(10) doesn't seem like a huge concern. I'll look at optimizing it but it seems doubtful it will have any actual impact.

Reply:
it's definitely not actually important!